src/github.com/matrix-org/dendrite/roomserver/input/input.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package input contains the code processes new room events
+package input
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+
+	"net/http"
+
+	"github.com/matrix-org/dendrite/common"
+	"github.com/matrix-org/dendrite/roomserver/api"
+	"github.com/matrix-org/util"
+	sarama "gopkg.in/Shopify/sarama.v1"
+)
+
+// RoomserverInputAPI implements api.RoomserverInputAPI
+type RoomserverInputAPI struct {
+	DB       RoomEventDatabase
+	Producer sarama.SyncProducer
+	// The kafkaesque topic to output new room events to.
+	// This is the name used in kafka to identify the stream to write events to.
+	OutputRoomEventTopic string
+	// If non-nil then the API will stop processing messages after this
+	// many messages and will shutdown. Malformed messages are not in the count.
+	StopProcessingAfter *int64
+	// If not-nil then the API will call this to shutdown the server.
+	// If this is nil then the API will continue to process messsages even
+	// though StopProcessingAfter has been reached.
+	ShutdownCallback func(reason string)
+	// How many messages the consumer has processed.
+	processed int64
+}
+
+// WriteOutputRoomEvent implements OutputRoomEventWriter
+func (r *RoomserverInputAPI) WriteOutputRoomEvent(output api.OutputNewRoomEvent) error {
+	var m sarama.ProducerMessage
+	oe := api.OutputEvent{
+		Type:         api.OutputTypeNewRoomEvent,
+		NewRoomEvent: &output,
+	}
+	value, err := json.Marshal(oe)
+	if err != nil {
+		return err
+	}
+	m.Topic = r.OutputRoomEventTopic
+	m.Key = sarama.StringEncoder("")
+	m.Value = sarama.ByteEncoder(value)
+	_, _, err = r.Producer.SendMessage(&m)
+	return err
+}
+
+// InputRoomEvents implements api.RoomserverInputAPI
+func (r *RoomserverInputAPI) InputRoomEvents(
+	request *api.InputRoomEventsRequest,
+	response *api.InputRoomEventsResponse,
+) error {
+	for i := range request.InputRoomEvents {
+		if err := processRoomEvent(r.DB, r, request.InputRoomEvents[i]); err != nil {
```

Comment:
Well, I had to change the `c` to `r` and the shutdown handler is *slightly* different. But yes it is taken from a copypasta.